standardize TLS connection properties

jamescassell opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

There are are a variety of options for TLS connection options and settings.  These should be standardized:
- `client_cert`: certificate for client identity, might also include the private key
- `client_key`: private key for `client_cert`
- `ca_cert`: public key to validate server's identity, usually a root certificate
- `validate_certs`: boolean to enable or disable certificate validity checking

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
fixes https://github.com/ansible/ansible/issues/52506

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request
- Docs Pull Request
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

jamescassell committed:
openstack: standardize tls params

jamescassell committed:
tower: tower_verify_ssl->validate_certs

jamescassell commented:
I reverted the OpenStack change to avoid wasting test resources.  The test passed.  You can see that change here: https://github.com/ansible/ansible/commit/d3c8602d6d5ff61cefada8029759eea89b408375

I'll add it again at the end once I've made the

jamescassell committed:
docker: use standard tls config params

- cacert_path -> ca_cert
- cert_path -> client_cert
- key_path -> client_key
- tls_verify -> validate_certs

jamescassell committed:
k8s: standardize tls connection params

- verify_ssl -> validate_certs
- ssl_ca_cert -> ca_cert
- cert_file -> client_cert
- key_file -> client_key

jamescassell committed:
ingate: verify_ssl -> validate_certs

jamescassell committed:
manageiq: standardize tls params

- verify_ssl -> validate_certs
- ca_bundle_path -> ca_cert

jamescassell committed:
mysql: standardize tls params

- ssl_ca -> ca_cert
- ssl_cert -> client_cert
- ssl_key -> client_key

jamescassell committed:
nios: ssl_verify -> validate_certs

felixfontein commented:
@jamescassell can you also update the docker_swarm inventory plugin (https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/inventory/docker_swarm.py)? It should use the same configuration names as the docker_* modules. (CC @morph027)

willthames commented:
Looks reasonable to me.

jamescassell committed:
postgresql: ssl_rootcert -> ca_cert

jamescassell committed:
rabbitmq: standardize tls params

- cacert -> ca_cert
- cert -> client_cert
- key -> client_key

jamescassell commented:
@felixfontein I'm saving plugins for last.  Hoping to get at least the module changes done in time for 2.8.  I'll get it if I have time.  If the changes are obvious to you, fee free to push to my branch.

jamescassell committed:
rackspace: verify_ssl -> validate_certs

jamescassell committed:
vca: verify_certs -> validate_certs

jamescassell committed:
kubevirt_cdi_upload: upload_host_verify_ssl -> upload_host_validate_certs

jamescassell committed:
lxd: standardize tls params

- key_file -> client_key
- cert_file -> client_cert

mmazur commented:
For `kubevirt_cdi_upload` either drop your changes entirely or (probably better) rename the existing arg to `upload_host_validate_certs`. Otherwise you'll be breaking my code.

cben commented:
Excellent effort :clap:  
LGTM for manageiq part.

jamescassell commented:
@mmazur I didn't add an alias in that case because the module is new in 2.8, but I'll see about adding an alias as I have done for the others.

mmazur commented:
@jamescassell I don't follow. Your patch compresses two separate arguments: k8s–inherited `verify_ssl` and _kubevirt_cdi_upload_–specific `upload_host_verify_ssl` into a single `validate_certs`. The only acceptable change is:

- (k8s–inherited) `verify_ssl` –> `validate_certs`
- `upload_host_verify_ssl` –> `upload_host_validate_certs`

Those need to be two separate sets of settings, as they control two separate code paths connecting to two separate systems.

jamescassell commented:
@mmazur I understand and will fix.

jamescassell committed:
get_certificate: ca_certs -> ca_cert

MarkusTeufelberger commented on lib/ansible/modules/crypto/get_certificate.py:
```diff
@@ -24,11 +24,12 @@
-    ca_certs:
+    ca_cert:
       description:
         - A PEM file containing a list of root certificates; if present, the cert will be validated against these root certs.
```
```suggestion
         - A PEM file containing one or more root certificates; if present, the cert will be validated against these root certs.
```

Commit:
get_certificate.py: clarify one or more certs in a file

Co-Authored-By: jamescassell <code@james.cassell.me>

jamescassell committed:
zabbix: tls_issuer -> ca_cert

felixfontein commented on lib/ansible/modules/crypto/get_certificate.py:
```diff
@@ -24,11 +24,12 @@
+        - A PEM file containing one or more root certificates; if present, the cert will be validated against these root certs.
         - Note that this only validates the certificate is signed by the chain; not that the cert is valid for the host presenting it.
       type: path
+      aliases: [ ca_certs ]
```
Since this module just has been added for 2.8, how about removing the alias and just keeping `ca_cert`? @MarkusTeufelberger @john-westcott-iv

jamescassell replied:
That's even better as long as there are no objections.

felixfontein replied:
I'll create a follow-up PR for this once this is merged.

jamescassell committed:
bigip_device_auth_ldap: standardize tls params

- ssl_check_peer -> validate_certs
- ssl_client_cert -> client_cert
- ssl_client_key -> client_key
- ssl_ca_cert -> ca_cert

maxamillion commented:
Glad this is being done. 👍

jamescassell committed:
vdirect: vdirect_validate_certs -> validate_certs

jamescassell committed:
mqtt: standardize tls params

- ca_certs -> ca_cert
- certfile -> client_cert
- keyfile -> client_key

jamescassell committed:
pulp_repo: standardize tls params

remove `importer_ssl` prefix

jamescassell committed:
rhn_register: sslcacert -> ca_cert

jamescassell committed:
yum_repository: standardize tls params

The fix for yum_repository is not straightforward since this module is
only a thin wrapper for the underlying commands and config.  In this
case, we add the new values as aliases, keeping the old as primary,
only due to the internal structure of the module.

Aliases added:
- sslcacert -> ca_cert
- sslclientcert -> client_cert
- sslclientkey -> client_key
- sslverify -> validate_certs

jamescassell committed:
gitlab_hook: enable_ssl_verification -> hook_validate_certs

felixfontein committed:
Adjust arguments for docker_swarm inventory plugin.

jamescassell committed:
foreman callback: standardize tls params

- ssl_cert -> client_cert
- ssl_key -> client_key

jamescassell committed:
grafana_annotations: validate_grafana_certs -> validate_certs

jamescassell committed:
nrdp callback: validate_nrdp_certs -> validate_certs

jamescassell committed:
oc connection: standardize tls params

- oc_cert_file -> client_cert
- oc_key_file -> client_key
- oc_ssl_ca_cert -> ca_cert
- oc_verify_ssl -> validate_certs

jamescassell committed:
kubectl connection: standardize tls params

- kubectl_cert_file -> client_cert
- kubectl_key_file -> client_key
- kubectl_ssl_ca_cert -> ca_cert
- kubectl_verify_ssl -> validate_certs

jamescassell committed:
psrp connection: cert_trust_path -> ca_cert

TODO: cert_validation -> validate_certs (multi-valued vs bool)

jamescassell committed:
k8s inventory: standardize tls params

- cert_file -> client_cert
- key_file -> client_key
- ca_cert -> ca_cert
- verify_ssl -> validate_certs

jamescassell committed:
openshift inventory: standardize tls params

- cert_file -> client_cert
- key_file -> client_key
- ca_cert -> ca_cert
- verify_ssl -> validate_certs

jamescassell committed:
tower inventory: verify_ssl -> validate_certs

jamescassell committed:
hashi_vault lookup: cacert -> ca_cert

jamescassell committed:
k8s lookup: standardize tls params

- cert_file -> client_cert
- key_file -> client_key
- ca_cert -> ca_cert
- verify_ssl -> validate_certs

jamescassell committed:
laps_passord lookup: cacert_file -> ca_cert

jamescassell committed:
changelog for TLS parameter standardization

jamescassell commented:
ready_for_review

felixfontein commented on lib/ansible/modules/packaging/os/yum_repository.py:
```diff
@@ -273,19 +273,23 @@
     description:
       - Path to the directory containing the databases of the certificate
         authorities yum should use to verify SSL certificates.
+    aliases: [ ca_cert ]
```
Why keeping the old names here, and adding the new ones only as an alias?

jamescassell replied:
the implementation hard-codes the internal option name to match the name in the yum repo file... it's a thin wrapper.  Detailed in the commit message: https://github.com/ansible/ansible/pull/54315/commits/053922ccb9e55c053f2cd6d660bc276125085513

felixfontein reviewed (approved):
shipit

mmazur commented:
+1  on cloud/kubevirt changes

maxamillion commented:
This is spectacular, thank you for all the work. 👍 

rebuild_merge

felixfontein commented:
@jamescassell thanks for doing the main work for this!
And thanks to everyeone who reviewed parts of this!
@maxamillion thanks for merging!

Andersson007 commented:
Guys, please, the next time take a short note to https://github.com/ansible/community/issues/435

wheerd commented:
I believe this change breaks the SSL certificate verification on the k8s module, as the parameters get passed down all the way to this REST Client: https://github.com/kubernetes-client/python/blob/master/kubernetes/client/rest.py
When using `ssl_ca_cert` and switching from 2.7.12 to 2.8.2 we get an SSL handshake error.
Should I open another issue for this?

felixfontein commented:
@wheerd you should open a new issue for this.
